File: sdk/go/provider/tempmailten.go
package provider

import (
	"encoding/json"
	"fmt"
	"io"
	"net/url"
	"regexp"
	"strings"

	http "github.com/bogdanfinn/fhttp"
)

/* tempmailten.com 临时邮箱服务商（Laravel + CSRF）
 * 流程：GET /en 获取 session cookie（cookie jar 自动维护）+ 从 HTML meta 提取 CSRF token
 *       POST /messages（body: _token={csrf}&captcha=）获取邮箱地址和邮件列表
 *       GET /view/{id} 获取单封邮件 HTML 正文
 * token 存储 CSRF token 值，session 由 tls-client 的 cookie jar 自动维护
 * 该站点 reCAPTCHA 已禁用（check_recaptcha 直接设为 true），captcha 参数传空即可
 */

const tempmailtenBaseURL = "https://tempmailten.com"

/* tempmailtenCSRFRe 从 meta 标签提取 CSRF token（<meta name="csrf-token" content="xxx">） */
var tempmailtenCSRFRe = regexp.MustCompile(`<meta\s+name="csrf-token"\s+content="([^"]+)"`)

/* tempmailtenBrowserHeaders 设置浏览器模拟请求头 */
func tempmailtenBrowserHeaders(req *http.Request) {
	req.Header.Set("User-Agent", getCurrentUA())
	req.Header.Set("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8")
	req.Header.Set("Accept-Language", "en-US,en;q=0.9,zh-CN;q=0.8,zh;q=0.7")
}

/* tempmailtenAjaxHeaders 设置 AJAX 请求头 */
func tempmailtenAjaxHeaders(req *http.Request) {
	req.Header.Set("User-Agent", getCurrentUA())
	req.Header.Set("Accept", "application/json, text/javascript, */*; q=0.01")
	req.Header.Set("Accept-Language", "en-US,en;q=0.9,zh-CN;q=0.8,zh;q=0.7")
	req.Header.Set("X-Requested-With", "XMLHttpRequest")
	req.Header.Set("Referer", tempmailtenBaseURL+"/en")
}

/* TempmailtenGenerate 创建 tempmailten.com 临时邮箱
 * 流程:
 *   1. GET /en 获取 session cookie 和 CSRF token（正则匹配 meta csrf-token）
 *   2. POST /messages（body: _token={csrf}&captcha=）获取分配的邮箱地址
 * token: CSRF token 值
 */
func TempmailtenGenerate() (*CreatedMailbox, error) {
	client := HTTPClient()

	req, err := http.NewRequest("GET", tempmailtenBaseURL+"/en", nil)
	if err != nil {
		return nil, fmt.Errorf("tempmailten: 创建首页请求失败: %w", err)
	}
	tempmailtenBrowserHeaders(req)

	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("tempmailten: 获取首页失败: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("tempmailten: 读取首页响应失败: %w", err)
	}

	if err := CheckHTTPStatus(resp, "tempmailten home"); err != nil {
		return nil, err
	}

	m := tempmailtenCSRFRe.FindSubmatch(body)
	if m == nil {
		return nil, fmt.Errorf("tempmailten: 未能从首页提取 CSRF token")
	}
	csrfToken := string(m[1])

	form := url.Values{}
	form.Set("_token", csrfToken)
	form.Set("captcha", "")

	req2, err := http.NewRequest("POST", tempmailtenBaseURL+"/messages", strings.NewReader(form.Encode()))
	if err != nil {
		return nil, fmt.Errorf("tempmailten: 创建消息请求失败: %w", err)
	}
	tempmailtenAjaxHeaders(req2)
	req2.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req2.Header.Set("X-CSRF-TOKEN", csrfToken)

	resp2, err := client.Do(req2)
	if err != nil {
		return nil, fmt.Errorf("tempmailten: 获取邮箱失败: %w", err)
	}
	defer resp2.Body.Close()

	body2, err := io.ReadAll(resp2.Body)
	if err != nil {
		return nil, fmt.Errorf("tempmailten: 读取邮箱响应失败: %w", err)
	}

	if err := CheckHTTPStatus(resp2, "tempmailten messages"); err != nil {
		return nil, err
	}

	emailAddr, err := tempmailtenExtractMailbox(body2)
	if err != nil {
		return nil, err
	}

	return &CreatedMailbox{
		Channel: "tempmailten",
		Email:   emailAddr,
		Token:   csrfToken,
	}, nil
}

/* TempmailtenGetEmails 获取 tempmailten.com 邮件列表
 * 流程:
 *   1. POST /messages（body: _token={csrf}&captcha=）获取 {mailbox, messages} JSON
 *   2. 对每封邮件 GET /view/{id} 获取 HTML 正文
 * messages 数组每个元素: {id, from, from_email, subject, is_seen}
 */
func TempmailtenGetEmails(email, token string) ([]NormEmail, error) {
	email = strings.TrimSpace(email)
	if email == "" {
		return nil, fmt.Errorf("tempmailten: 邮箱地址为空")
	}
	csrfToken := strings.TrimSpace(token)
	if csrfToken == "" {
		return nil, fmt.Errorf("tempmailten: 缺少 CSRF token")
	}

	client := HTTPClient()

	form := url.Values{}
	form.Set("_token", csrfToken)
	form.Set("captcha", "")

	req, err := http.NewRequest("POST", tempmailtenBaseURL+"/messages", strings.NewReader(form.Encode()))
	if err != nil {
		return nil, fmt.Errorf("tempmailten: 创建消息请求失败: %w", err)
	}
	tempmailtenAjaxHeaders(req)
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("X-CSRF-TOKEN", csrfToken)

	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("tempmailten: 获取邮件列表失败: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("tempmailten: 读取邮件列表响应失败: %w", err)
	}

	if err := CheckHTTPStatus(resp, "tempmailten messages"); err != nil {
		return nil, err
	}

	return tempmailtenParseMessages(body, email, csrfToken, client)
}

/* tempmailtenExtractMailbox 从 POST /messages 的 JSON 响应中提取邮箱地址 */
func tempmailtenExtractMailbox(data []byte) (string, error) {
	parsed := tempmailtenParseJSON(data)
	if parsed == nil {
		return "", fmt.Errorf("tempmailten: 解析响应 JSON 失败")
	}

	mailbox, ok := parsed["mailbox"].(string)
	if !ok || mailbox == "" {
		return "", fmt.Errorf("tempmailten: 响应中未包含有效的 mailbox 字段")
	}

	mailbox = strings.TrimSpace(mailbox)
	if !strings.Contains(mailbox, "@") {
		return "", fmt.Errorf("tempmailten: 邮箱地址无效: %q", mailbox)
	}

	return mailbox, nil
}

/* tempmailtenParseMessages 解析邮件列表并拉取正文 */
func tempmailtenParseMessages(data []byte, email, csrfToken string, client interface {
	Do(*http.Request) (*http.Response, error)
}) ([]NormEmail, error) {
	parsed := tempmailtenParseJSON(data)
	if parsed == nil {
		return nil, fmt.Errorf("tempmailten: 解析响应 JSON 失败")
	}

	msgsRaw, ok := parsed["messages"].([]interface{})
	if !ok || len(msgsRaw) == 0 {
		return []NormEmail{}, nil
	}

	emails := make([]NormEmail, 0, len(msgsRaw))
	for _, raw := range msgsRaw {
		msg, ok := raw.(map[string]interface{})
		if !ok {
			continue
		}

		id := tempmailtenStr(msg["id"])
		if id == "" {
			continue
		}

		fromAddr := tempmailtenStr(msg["from_email"])
		fromName := tempmailtenStr(msg["from"])
		if fromName != "" && fromName != fromAddr {
			fromAddr = fmt.Sprintf("%s <%s>", fromName, fromAddr)
		}

		htmlBody := tempmailtenFetchView(client, id)

		isSeen, _ := msg["is_seen"].(bool)

		flat := map[string]interface{}{
			"id":      id,
			"from":    fromAddr,
			"to":      email,
			"subject": tempmailtenStr(msg["subject"]),
			"html":    htmlBody,
			"isRead":  isSeen,
		}
		emails = append(emails, NormalizeMap(flat, email))
	}

	return emails, nil
}

/* tempmailtenFetchView 获取单封邮件的 HTML 正文（GET /view/{id}） */
func tempmailtenFetchView(client interface {
	Do(*http.Request) (*http.Response, error)
}, id string) string {
	if id == "" {
		return ""
	}

	req, err := http.NewRequest("GET", tempmailtenBaseURL+"/view/"+url.PathEscape(id), nil)
	if err != nil {
		return ""
	}
	tempmailtenBrowserHeaders(req)
	req.Header.Set("Referer", tempmailtenBaseURL+"/en")

	resp, err := client.Do(req)
	if err != nil {
		return ""
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		io.Copy(io.Discard, resp.Body)
		return ""
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return ""
	}

	return string(body)
}

/* tempmailtenParseJSON 解析 JSON 为 map */
func tempmailtenParseJSON(data []byte) map[string]interface{} {
	var result map[string]interface{}
	if err := json.Unmarshal(data, &result); err != nil {
		return nil
	}
	return result
}

/* tempmailtenStr 安全提取 interface{} 为 string */
func tempmailtenStr(v interface{}) string {
	if v == nil {
		return ""
	}
	switch s := v.(type) {
	case string:
		return strings.TrimSpace(s)
	case float64:
		return fmt.Sprintf("%.0f", s)
	default:
		return fmt.Sprintf("%v", s)
	}
}

File: sdk/rust/src/providers/tempmailten.rs
/*!
 * tempmailten 渠道实现
 * 网站: https://tempmailten.com
 *
 * Laravel + CSRF 后端。
 * 流程:
 *   1. GET /en 获取 session cookie + 从 HTML meta 提取 CSRF token
 *      （<meta name="csrf-token" content="xxx">）
 *   2. POST /messages（body: _token={csrf}&captcha=）获取 {mailbox, messages}
 *   3. GET /view/{id} 获取单封邮件 HTML 正文
 *
 * token 存储: JSON 字符串 {"csrf":"...","cookie":"..."}
 * 该站点 reCAPTCHA 已禁用，captcha 参数传空即可。
 */

use std::collections::BTreeMap;
use std::sync::LazyLock;

use regex::Regex;
use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::config::{block_on, get_current_ua, http_client_no_cookie_jar};
use crate::normalize::normalize_email;
use crate::types::{Channel, Email, EmailInfo};

const BASE_URL: &str = "https://tempmailten.com";

/// 从 meta 标签提取 CSRF token（<meta name="csrf-token" content="xxx">）
static CSRF_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r#"<meta\s+name="csrf-token"\s+content="([^"]+)""#).expect("re"));

/// token 中序列化的会话信息
#[derive(Serialize, Deserialize)]
struct TempmailtenSess {
    /// CSRF token
    csrf: String,
    /// Cookie 字符串
    cookie: String,
}

/// 解析 Cookie 头为 key=value 映射
fn parse_cookie_header(hdr: &str) -> BTreeMap<String, String> {
    let mut m = BTreeMap::new();
    for part in hdr.split(';') {
        let part = part.trim();
        if part.is_empty() {
            continue;
        }
        if let Some(i) = part.find('=') {
            let k = part[..i].trim();
            let v = part[i + 1..].trim();
            if !k.is_empty() {
                m.insert(k.to_string(), v.to_string());
            }
        }
    }
    m
}

/// 将 Cookie 映射转换回 Cookie 头字符串
fn cookie_header_from_map(m: &BTreeMap<String, String>) -> String {
    m.iter()
        .map(|(k, v)| format!("{}={}", k, v))
        .collect::<Vec<_>>()
        .join("; ")
}

/// 合并 Set-Cookie 响应头到已有 Cookie 中
fn merge_set_cookies(hdr: &str, headers: &wreq::header::HeaderMap) -> String {
    let mut m = parse_cookie_header(hdr);
    for val in headers.get_all("set-cookie") {
        let Ok(s) = val.to_str() else {
            continue;
        };
        let first = s.split(';').next().unwrap_or("");
        if let Some(i) = first.find('=') {
            let k = first[..i].trim();
            let v = first[i + 1..].trim();
            if !k.is_empty() {
                m.insert(k.to_string(), v.to_string());
            }
        }
    }
    cookie_header_from_map(&m)
}

/// 编码会话信息为 token 字符串
fn encode_sess(s: &TempmailtenSess) -> Result<String, String> {
    serde_json::to_string(s).map_err(|e| format!("tempmailten: token 序列化失败: {}", e))
}

/// 从 token 字符串解码会话信息
fn decode_sess(tok: &str) -> Result<TempmailtenSess, String> {
    let s: TempmailtenSess =
        serde_json::from_str(tok).map_err(|_| "tempmailten: token 反序列化失败".to_string())?;
    if s.csrf.is_empty() {
        return Err("tempmailten: token 中 csrf 为空".into());
    }
    Ok(s)
}

/// 构造 POST /messages 的表单 body（_token={csrf}&captcha=）
fn messages_form(csrf: &str) -> String {
    format!("_token={}&captcha=", urlencoding::encode(csrf))
}

/// POST /messages 获取 {mailbox, messages} JSON
/// 返回解析后的 JSON 与合并后的 Cookie 串
async fn post_messages(
    client: &wreq::Client,
    csrf: &str,
    cookie_hdr: &str,
) -> Result<(Value, String), String> {
    let resp = client
        .post(format!("{BASE_URL}/messages"))
        .header("User-Agent", get_current_ua())
        .header("Accept", "application/json, text/javascript, */*; q=0.01")
        .header("Accept-Language", "en-US,en;q=0.9,zh-CN;q=0.8,zh;q=0.7")
        .header("X-Requested-With", "XMLHttpRequest")
        .header("Referer", format!("{BASE_URL}/en"))
        .header("Content-Type", "application/x-www-form-urlencoded")
        .header("X-CSRF-TOKEN", csrf)
        .header("Cookie", cookie_hdr)
        .body(messages_form(csrf))
        .send()
        .await
        .map_err(|e| format!("tempmailten: 请求 /messages 失败: {}", e))?;

    if !resp.status().is_success() {
        return Err(format!("tempmailten messages: {}", resp.status()));
    }

    let merged = merge_set_cookies(cookie_hdr, resp.headers());
    let text = resp
        .text()
        .await
        .map_err(|e| format!("tempmailten: 读取 /messages 响应失败: {}", e))?;
    let data: Value = serde_json::from_str(text.trim())
        .map_err(|e| format!("tempmailten: 解析 /messages 响应失败: {}", e))?;
    Ok((data, merged))
}

/// 创建临时邮箱
/// 1. GET /en 获取 session cookie 和 CSRF token
/// 2. POST /messages 获取分配的邮箱地址
pub fn generate_email() -> Result<EmailInfo, String> {
    block_on(async {
        let client = http_client_no_cookie_jar();

        // 第一步：GET /en 获取 session cookie 和 CSRF token
        let resp = client
            .get(format!("{BASE_URL}/en"))
            .header("User-Agent", get_current_ua())
            .header(
                "Accept",
                "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8",
            )
            .header("Accept-Language", "en-US,en;q=0.9,zh-CN;q=0.8,zh;q=0.7")
            .send()
            .await
            .map_err(|e| format!("tempmailten: 获取首页失败: {}", e))?;

        if !resp.status().is_success() {
            return Err(format!("tempmailten: 首页返回 HTTP {}", resp.status()));
        }

        let cookie_hdr = merge_set_cookies("", resp.headers());
        let html = resp
            .text()
            .await
            .map_err(|e| format!("tempmailten: 读取首页响应失败: {}", e))?;

        let csrf = CSRF_RE
            .captures(&html)
            .and_then(|c| c.get(1))
            .map(|m| m.as_str().to_string())
            .filter(|s| !s.is_empty())
            .ok_or_else(|| "tempmailten: 未能从首页提取 CSRF token".to_string())?;

        // 第二步：POST /messages 获取邮箱地址
        let (data, cookie_hdr) = post_messages(&client, &csrf, &cookie_hdr).await?;

        let mailbox = data
            .get("mailbox")
            .and_then(|v| v.as_str())
            .map(|s| s.trim().to_string())
            .unwrap_or_default();

        if mailbox.is_empty() || !mailbox.contains('@') {
            return Err("tempmailten: 响应中未包含有效的 mailbox 字段".into());
        }

        let tok = encode_sess(&TempmailtenSess {
            csrf,
            cookie: cookie_hdr,
        })?;

        Ok(EmailInfo {
            channel: Channel::Tempmailten,
            email: mailbox,
            token: Some(tok),
            expires_at: None,
            created_at: None,
        })
    })
}

/// 获取邮件列表
/// 1. POST /messages 获取 {mailbox, messages} JSON
/// 2. 对每封邮件 GET /view/{id} 获取 HTML 正文
pub fn get_emails(token: &str, email: &str) -> Result<Vec<Email>, String> {
    let em = email.trim();
    if em.is_empty() {
        return Err("tempmailten: 邮箱地址为空".into());
    }
    let sess = decode_sess(token)?;

    block_on(async {
        let client = http_client_no_cookie_jar();
        let (data, _) = post_messages(&client, &sess.csrf, &sess.cookie).await?;

        // messages 数组每个元素: {id, from, from_email, subject, is_seen}
        let msgs = match data.get("messages").and_then(|v| v.as_array()) {
            Some(arr) if !arr.is_empty() => arr,
            _ => return Ok(vec![]),
        };

        let mut result = Vec::with_capacity(msgs.len());
        for msg in msgs {
            let id = json_str(msg.get("id"));
            if id.is_empty() {
                continue;
            }

            let from_email = json_str(msg.get("from_email"));
            let from_name = json_str(msg.get("from"));
            let from_addr = if !from_name.is_empty() && from_name != from_email {
                format!("{} <{}>", from_name, from_email)
            } else {
                from_email
            };

            let is_seen = msg
                .get("is_seen")
                .and_then(|v| v.as_bool())
                .unwrap_or(false);
            let html_body = fetch_view(&client, &sess.cookie, &id).await;

            let raw = serde_json::json!({
                "id": id,
                "from": from_addr,
                "to": em,
                "subject": json_str(msg.get("subject")),
                "html": html_body,
                "is_read": is_seen,
            });
            result.push(normalize_email(&raw, em));
        }

        Ok(result)
    })
}

/// 获取单封邮件的 HTML 正文（GET /view/{id}）
async fn fetch_view(client: &wreq::Client, cookie_hdr: &str, id: &str) -> String {
    if id.is_empty() {
        return String::new();
    }

    let url = format!("{BASE_URL}/view/{}", urlencoding::encode(id));
    let resp = match client
        .get(&url)
        .header("User-Agent", get_current_ua())
        .header(
            "Accept",
            "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8",
        )
        .header("Accept-Language", "en-US,en;q=0.9,zh-CN;q=0.8,zh;q=0.7")
        .header("Referer", format!("{BASE_URL}/en"))
        .header("Cookie", cookie_hdr)
        .send()
        .await
    {
        Ok(r) if r.status().is_success() => r,
        _ => return String::new(),
    };

    resp.text().await.unwrap_or_default()
}

/// 安全提取 JSON 值为字符串（字符串原样、数字转整数文本）
fn json_str(v: Option<&Value>) -> String {
    match v {
        Some(v) if v.is_string() => v.as_str().unwrap_or("").trim().to_string(),
        Some(v) if v.is_number() => {
            if let Some(n) = v.as_i64() {
                n.to_string()
            } else if let Some(n) = v.as_u64() {
                n.to_string()
            } else {
                v.to_string()
            }
        }
        _ => String::new(),
    }
}
